src/common.cc

```diff
@@ -111,4 +121,337 @@ Join(const std::vector<std::string>& vec, const std::string& delim)
+/// 0 for the initial next_offset. Subsequent calls should pass the returned
+/// values without modification.
+///
+/// \param[in] datatype data type of array
+/// \param[in, out] base pointer to array of data type elements to convert
+/// \param[in] byte_size size of array in bytes
+/// \param[in] host_byte_order whether original array is in host byte
+/// order
+/// \param[in, out] partial_result vector to store partial results
+/// \param[in, out] next_offset intermediate value used for storing next offset
+/// for BYTES data type
+///
+void
+SwapEndian(
+    TRITONSERVER_DataType datatype, char* base, size_t byte_size,
+    bool host_byte_order, std::vector<char*>& partial_result,
+    size_t& next_offset)
+{
+  switch (datatype) {
+    case TRITONSERVER_TYPE_UINT16:
+    case TRITONSERVER_TYPE_INT16:
+    case TRITONSERVER_TYPE_FP16: {
+      SwapEndian<uint16_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_UINT32:
+    case TRITONSERVER_TYPE_INT32:
+    case TRITONSERVER_TYPE_FP32: {
+      SwapEndian<uint32_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_UINT64:
+    case TRITONSERVER_TYPE_INT64:
+    case TRITONSERVER_TYPE_FP64: {
+      SwapEndian<uint64_t>(base, byte_size, partial_result);
+      break;
+    }
+    case TRITONSERVER_TYPE_BYTES: {
+      uint32_t host_length = 0;
+      while (next_offset < byte_size) {
+        size_t partial_offset = partial_result.size();
+        bswap_length_partial(
+            base + next_offset, byte_size - next_offset, host_byte_order,
+            &host_length, partial_result);
+        if (partial_result.size() == 0) {
+          next_offset += sizeof(uint32_t) - partial_offset + host_length;
+        } else {
+          next_offset = 0;
+          return;
+        }
+      }
+      next_offset -= byte_size;
+      break;
+    }
+    default: {
+    }
+  }
+}
+
+void
+HostToLittleEndian(TRITONSERVER_DataType datatype, char* base, size_t byte_size)
+{
+  SwapEndian(datatype, base, byte_size, true);
+}
+
+void
+LittleEndianToHost(TRITONSERVER_DataType datatype, char* base, size_t byte_size)
+{
+  SwapEndian(datatype, base, byte_size, false);
+}
+
+void
+LittleEndianToHost(
+    TRITONSERVER_DataType datatype, char* base, size_t byte_size,
+    std::vector<char*>& partial_result, size_t& next_offset)
+{
+  SwapEndian(datatype, base, byte_size, false, partial_result, next_offset);
+}
+
+TRITONSERVER_DataType
+GetDataTypeForRawInput(
+    TRITONSERVER_Server* server, const std::string& model_name,
+    const int64_t model_version)
+{
+  TRITONSERVER_Message* message = nullptr;
+  triton::common::TritonJson::Value document;
+  triton::common::TritonJson::Value inputs;
+  triton::common::TritonJson::Value input;
+  const char* datatype = "";
+  size_t datatype_length = 0;
+  const char* buffer;
+  size_t byte_size;
+
+  RETURN_VAL_IF_ERR(
+      TRITONSERVER_ServerModelMetadata(
```

Comment:
I think this call is very costly. Is there a way to avoid making this call for every inference request in raw input case?

Reply:
I agree - we could cache the data types? but if models change that becomes an issue as well. Unfortunately in this case the request itself doesn't have enough information to know the data type. maybe we start here and think of how to optimize?

Reply:
I agree we can optimize this later since it only affects the performance of raw data on big endian. Resolving.

Reply:
Should we add a `NOTE/FIXME` type comment above? I don't think it would be obvious to future readers that this is costly without really staring at it.